File: algorithms/cpp/081_search_in_rotated_sorted_array_II.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        if (nums.empty()) return false;
        else return divide(nums,target,0, nums.size()-1);
    }
    int divide(vector<int>& nums, int target, int left, int right) {
        if (left>right)
            return false;
        int mid=(left+right)/2;
        if (target==nums[mid]) return true;
        if (nums[left]<nums[mid])  {
            if (target<=nums[mid] && target>=nums[left]) return divide(nums,target,left, mid-1);
            else return divide(nums,target,mid+1,right);
        }
        else if (nums[left]>nums[mid]) {
            if (target>nums[mid] && target<=nums[right]) return divide(nums,target, mid+1,right);
            else return divide(nums,target,left, mid-1);
        }
        else return divide(nums,target,left+1,right);
    }
};

File: algorithms/python3/081_search_in_rotated_sorted_array_II.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        def helper(nums,start,end,target):
            if start>end:
                return False
            mid=(start+end)//2
            if nums[mid]==target:
                return True
            if nums[mid]>nums[start]:
                if nums[mid]>target and target>=nums[start]:
                    return helper(nums,start,mid-1,target)
                else:
                    return helper(nums,mid+1,end,target)

            elif nums[mid]<nums[start]:
                if nums[mid]<target and target<=nums[end]:
                    return helper(nums,mid+1,end,target)
                else:
                    return helper(nums,start,mid-1,target)
            else:
                return helper(nums,start+1,end,target)
        return helper(nums,0,len(nums)-1,target)


a=Solution()

nums=[2,5,6,0,0,1,2]
a.search(nums,3)

